### PR TITLE
Focus issue in editor settings menu #3921

### PR DESCRIPTION
**Pull request by jjjoness**

### CODE REVIEW

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ AZ_PUSH_DISABLE_DLL_EXPORT_MEMBER_WARNING
 AZ_POP_DISABLE_DLL_EXPORT_MEMBER_WARNING
 
 using namespace AzQtComponents;
-
+#pragma optimize("",off)
```
Comment (AMZN-daimini) [inline Code/Editor/EditorPreferencesDialog.cpp:42]: This should be removed before submission

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 2]
```diff
@@ -112,6 +112,24 @@ void EditorPreferencesDialog::showEvent(QShowEvent* event)
     QDialog::showEvent(event);
 }
 
+QList<const char*> list;
```
Comment (AMZN-daimini) [inline Code/Editor/EditorPreferencesDialog.cpp:115]: Should this be here?

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 3]
```diff
@@ -112,6 +112,24 @@ void EditorPreferencesDialog::showEvent(QShowEvent* event)
...
+    {
+        QDialog::keyPressEvent(event);
+    }
+    else
+    {
+        QWidget* widget = QApplication::focusWidget();
+        if (widget)
+        {
+            widget->clearFocus();
+        }
```
Comment (AMZN-daimini) [inline Code/Editor/EditorPreferencesDialog.cpp:129]: nit: you can convert this to
```
if (QWidget* widget = QApplication::focusWidget(); widget != nullptr)
{
        widget->clearFocus();
}
```

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 4]
```diff
@@ -112,6 +112,24 @@ void EditorPreferencesDialog::showEvent(QShowEvent* event)
     QDialog::showEvent(event);
 }
 
+QList<const char*> list;
+
+void EditorPreferencesDialog::keyPressEvent([[maybe_unused]]QKeyEvent* event)
```
Comment (AMZN-daimini) [inline Code/Editor/EditorPreferencesDialog.cpp:117]: I'm not entirely sure this is the proper way to handle the issue. The intention is to replace the preference system though, so I'd be ok with this if it's meant to be temporary.

Else, I think the input should consume the event and prevent it from being propagated out.

File: Code/Editor/EditorPreferencesDialog.cpp (hunk 1 above)
Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.cpp:42]: Oopps, thought I'd checked! Done.

File: Code/Editor/EditorPreferencesDialog.cpp (hunk 2 above)
Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.cpp:115]: Again oops. Removed.

File: Code/Editor/EditorPreferencesDialog.cpp (hunk 4 above)
Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.cpp:117]: keyPressEvent assumes it is consumed unless you explicitly call the parent.

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 5]
```diff
@@ -112,6 +112,22 @@ void EditorPreferencesDialog::showEvent(QShowEvent* event)
     QDialog::showEvent(event);
 }
 
+
+void EditorPreferencesDialog::keyPressEvent(QKeyEvent* event)
```
Comment (hultonha) [inline Code/Editor/EditorPreferencesDialog.cpp:116]: nit: line 124 the formatting of the if is a bit off (should be `if (QWidget*...`) - if you select this block and hit Ctrl-K, F then clang-format will do its thing and auto-format it all

File: Code/Editor/EditorPreferencesDialog.cpp [hunk 6]
```diff
@@ -112,6 +112,22 @@ void EditorPreferencesDialog::showEvent(QShowEvent* event)
...
+{
+    if (event->key() != Qt::Key::Key_Enter && event->key() != Qt::Key::Key_Return)
+    {
+        QDialog::keyPressEvent(event);
+    }
+    else
+    {
+        if( QWidget* widget = QApplication::focusWidget(); widget != nullptr)
```
Comment (hultonha) [inline Code/Editor/EditorPreferencesDialog.cpp:124]: nit: This can be simplified to..

```
if (QWidget* widget = QApplication::focusWidget())
```

**Review by hultonha — APPROVED** [review]
The change looks good but it would be great if we could create a test to validate this behavior. We could do this my moving this code to a helper free function, that we can call from a simple widget from a test fixture (instead of having to try and instantiate `EditorPreferencesDialog`). We can then check that the event is consumed correctly and doesn't make it to the next parent widget for example. This could be a separate PR but it would definitely be worth spending a bit of time investigating how to do it.

File: Code/Editor/EditorPreferencesDialog.cpp (hunk 5 above)
Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.cpp:116]: Re-formatted.

File: Code/Editor/EditorPreferencesDialog.cpp (hunk 6 above)
Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.cpp:124]: Nice.

File: Code/Editor/EditorPreferencesDialog.h [hunk 7]
```diff
@@ -18,6 +18,26 @@ namespace Ui
 }
 
 class EditorPreferencesTreeWidgetItem;
+void WidgetHandleKeyPressEvent(QWidget* widget, QKeyEvent* event)
+{
+    // If the enter key is pressed during any text input, the dialog box will close
+    // making it inconvenient to do multiple edits. This routine captures the
+    // Key_Enter or Key_Return and clears the focus to give a visible cue that
+    // editing of that field has finished and then doesn't propogate it.
```
Comment (hultonha) [inline Code/Editor/EditorPreferencesDialog.h:26]: Great comment! Thanks for adding 😄

File: Code/Editor/EditorPreferencesDialog.h [hunk 8]
```diff
@@ -18,6 +18,26 @@ namespace Ui
 }
 
 class EditorPreferencesTreeWidgetItem;
+void WidgetHandleKeyPressEvent(QWidget* widget, QKeyEvent* event)
```
Comment (hultonha) [inline Code/Editor/EditorPreferencesDialog.h:21]: update: We should leave the definition in the .cpp file but keep the declaration in the .h - either as part of this PR or a follow-up we should then create a dummy widget to invoke this call in its own `keyPressEvent` and verify the behavior matches our expectations

Comment (jjjoness) [reply Code/Editor/EditorPreferencesDialog.h]: Changed.

**Review by hultonha — APPROVED** [review]
LGTM (just need to write the test now 😉)

**Review by AMZN-daimini — APPROVED** [review]

**Pull request merged**